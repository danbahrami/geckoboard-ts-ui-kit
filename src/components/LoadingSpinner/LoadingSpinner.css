.ui-loading-spinner {
  position: relative;
  margin-right: auto;
  margin-left: auto;
  width: 22px;
  height: 22px;
  animation: smoothLoader;
}

.ui-loading-spinner__circle {
  width: 100%;
  height: 100%;
  position: absolute;
  left: 0;
  top: 0;
}

.ui-loading-spinner__circle--theme-dark .ui-loading-spinner__circle__inner {
  background-color: #4a4a4a;
}

.ui-loading-spinner__circle__inner {
  display: block;
  margin: 0 auto;
  background-color: #dadada;
  animation: sk-circleFadeDelay 1.2s ease-in-out infinite;
  animation-fill-mode: both;
  border-radius: 50%;
}

.ui-loading-spinner__circle-2 {
  transform: rotate(30deg);
}
.ui-loading-spinner__circle-2 .ui-loading-spinner__circle__inner {
  animation-delay: -1.1s;
}

.ui-loading-spinner__circle-3 {
  transform: rotate(60deg);
}
.ui-loading-spinner__circle-3 .ui-loading-spinner__circle__inner {
  animation-delay: -1s;
}

.ui-loading-spinner__circle-4 {
  transform: rotate(90deg);
}
.ui-loading-spinner__circle-4 .ui-loading-spinner__circle__inner {
  animation-delay: -0.9s;
}

.ui-loading-spinner__circle-5 {
  transform: rotate(120deg);
}
.ui-loading-spinner__circle-5 .ui-loading-spinner__circle__inner {
  animation-delay: -0.8s;
}

.ui-loading-spinner__circle-6 {
  transform: rotate(150deg);
}
.ui-loading-spinner__circle-6 .ui-loading-spinner__circle__inner {
  animation-delay: -0.7s;
}

.ui-loading-spinner__circle-7 {
  transform: rotate(180deg);
}
.ui-loading-spinner__circle-7 .ui-loading-spinner__circle__inner {
  animation-delay: -0.6s;
}

.ui-loading-spinner__circle-8 {
  transform: rotate(210deg);
}
.ui-loading-spinner__circle-8 .ui-loading-spinner__circle__inner {
  animation-delay: -0.5s;
}

.ui-loading-spinner__circle-9 {
  transform: rotate(240deg);
}
.ui-loading-spinner__circle-9 .ui-loading-spinner__circle__inner {
  animation-delay: -0.4s;
}

.ui-loading-spinner__circle-10 {
  transform: rotate(270deg);
}
.ui-loading-spinner__circle-10 .ui-loading-spinner__circle__inner {
  animation-delay: -0.3s;
}

.ui-loading-spinner__circle-11 {
  transform: rotate(300deg);
}
.ui-loading-spinner__circle-11 .ui-loading-spinner__circle__inner {
  animation-delay: -0.2s;
}

.ui-loading-spinner__circle-12 {
  transform: rotate(330deg);
}
.ui-loading-spinner__circle-12 .ui-loading-spinner__circle__inner {
  animation-delay: -0.1s;
}

@keyframes sk-circleFadeDelay {
  0%,
  39%,
  100% {
    opacity: 0;
  }
  40% {
    opacity: 1;
  }
}

/* From https://github.com/tobiasahlin/SpinKit/blob/master/css/spinners/10-fading-circle.css */
@keyframes smoothLoader {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
