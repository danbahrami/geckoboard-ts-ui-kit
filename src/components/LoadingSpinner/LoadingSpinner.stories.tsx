import { Story, Meta } from '@storybook/react';

import { LoadingSpinner, LoadingSpinnerProps } from './LoadingSpinner';

export default {
  title: 'Components/Loading Spinner',
  component: LoadingSpinner,
  argTypes: {},
} as Meta<typeof LoadingSpinner>;

const Template: Story<LoadingSpinnerProps> = (args) => (
  <LoadingSpinner {...args} />
);

export const Basic = Template.bind({});
Basic.args = {};

export const Dark = Template.bind({});
Dark.args = {
  theme: 'dark',
};

export const Small = Template.bind({});
Small.args = {
  size: 16,
};

export const Large = Template.bind({});
Large.args = {
  size: 30,
};
