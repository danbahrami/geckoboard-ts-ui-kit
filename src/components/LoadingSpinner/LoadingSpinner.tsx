import React from 'react';
import range from 'lodash/range';
import round from 'lodash/round';
import classnames from 'classnames';

import './LoadingSpinner.css';

const defaultProps = {
  size: 20,
  theme: 'light',
};

export interface LoadingSpinnerProps {
  size: number;

  theme?: 'light' | 'dark';
}

// We should make sure that the circles always have a
// width and height that are a round number. Browsers
// struggle to give good border-radii to elements with
// non-integer dimensions.
const getCirclePxSize = (size: number) => round(size * 0.15);

export const LoadingSpinner = (
  props: LoadingSpinnerProps,
): React.ReactElement => {
  const { size = defaultProps.size, theme = defaultProps.theme } = props;

  const circlePxSize = getCirclePxSize(size);

  const renderCircle = (n: number) => {
    const classes = classnames(
      'ui-loading-spinner__circle',
      `ui-loading-spinner__circle-${n}`,
      {
        'ui-loading-spinner__circle--theme-dark': theme === 'dark',
      },
    );

    return (
      <div key={n} className={classes}>
        <span
          className="ui-loading-spinner__circle__inner"
          style={{
            width: `${circlePxSize}px`,
            height: `${circlePxSize}px`,
          }}
        />
      </div>
    );
  };

  return (
    <div className="ui-loading-spinner" style={{ width: size, height: size }}>
      {range(1, 13).map(renderCircle)}
    </div>
  );
};

LoadingSpinner.defaultProps = defaultProps;
