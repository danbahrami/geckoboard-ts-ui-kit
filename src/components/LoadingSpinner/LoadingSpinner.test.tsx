import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { LoadingSpinner } from './LoadingSpinner';

describe('Components / LoadingSpinner', () => {
  let container: Element;

  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
  });

  it('renders a loading spinner', () => {
    act(() => {
      render(<LoadingSpinner />, container);
    });

    expect(container).toMatchSnapshot();
  });

  it('lets you pick a size', () => {
    act(() => {
      render(<LoadingSpinner size={40} />, container);
    });

    expect(container).toMatchSnapshot();
  });

  it('lets you apply a `dark` theme', () => {
    act(() => {
      render(<LoadingSpinner theme="dark" />, container);
    });

    expect(container).toMatchSnapshot();
  });
});
