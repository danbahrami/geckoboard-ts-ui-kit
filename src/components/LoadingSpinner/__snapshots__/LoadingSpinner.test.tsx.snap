// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components / LoadingSpinner lets you apply a \`dark\` theme 1`] = `
<div>
  <div
    class="ui-loading-spinner"
    style="width: 20px; height: 20px;"
  >
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-1 ui-loading-spinner__circle--theme-dark"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-2 ui-loading-spinner__circle--theme-dark"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-3 ui-loading-spinner__circle--theme-dark"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-4 ui-loading-spinner__circle--theme-dark"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-5 ui-loading-spinner__circle--theme-dark"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-6 ui-loading-spinner__circle--theme-dark"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-7 ui-loading-spinner__circle--theme-dark"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-8 ui-loading-spinner__circle--theme-dark"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-9 ui-loading-spinner__circle--theme-dark"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-10 ui-loading-spinner__circle--theme-dark"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-11 ui-loading-spinner__circle--theme-dark"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-12 ui-loading-spinner__circle--theme-dark"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
  </div>
</div>
`;

exports[`Components / LoadingSpinner lets you pick a size 1`] = `
<div>
  <div
    class="ui-loading-spinner"
    style="width: 40px; height: 40px;"
  >
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-1"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 6px; height: 6px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-2"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 6px; height: 6px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-3"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 6px; height: 6px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-4"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 6px; height: 6px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-5"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 6px; height: 6px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-6"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 6px; height: 6px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-7"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 6px; height: 6px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-8"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 6px; height: 6px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-9"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 6px; height: 6px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-10"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 6px; height: 6px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-11"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 6px; height: 6px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-12"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 6px; height: 6px;"
      />
    </div>
  </div>
</div>
`;

exports[`Components / LoadingSpinner renders a loading spinner 1`] = `
<div>
  <div
    class="ui-loading-spinner"
    style="width: 20px; height: 20px;"
  >
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-1"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-2"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-3"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-4"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-5"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-6"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-7"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-8"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-9"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-10"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-11"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
    <div
      class="ui-loading-spinner__circle ui-loading-spinner__circle-12"
    >
      <span
        class="ui-loading-spinner__circle__inner"
        style="width: 3px; height: 3px;"
      />
    </div>
  </div>
</div>
`;
